### ISSUE TITLE
Using redis with rustls is blocking

### ISSUE THREAD

**Issue by NicoSchmidt1703** [redis-rs/redis-rs#1256]
Over the last few days @FalkWoldmann and me dug in a few performance issues on our instances. We want to share our results with you because it might be relevant for you too. 

First, let me start by giving you some insights on what our application does:
We have a controller with a hard-coded timeout, every request of a user results in a backend http call to a service which often exceeds this timeout. Therefore, we are forced to create a shared redis cache, which we fill whether the controller times out or not. 
We saw, that our newly created application cannot handle the number of requests we hoped it will. Which makes us suspect that there is something wrong. 

We found something interesting. But first, give me a bit of code, so you're able to replicate our findings. (My tests are on macOS, but our applications run on amazonlinux with the same issues)

```toml
[package]
name = "example"
version = "0.1.0"
edition = "2021"

[dependencies]
tower = { version = "0.4", features = ["util"] }
tower-http = { version = "0.5.2", features = ["timeout", "fs", "trace"] }
axum = { version = "0.7.5", features = ["macros"] }
tokio = { version = "1.38.0", features = ["full"] }
log = "0.4.22"
anyhow = "1.0.86"
chrono = "0.4.38"
axum-extra = { version = "0.9.3", features = ["cookie", "query"] }

redis = { version = "0.25.4", features = [
    "tokio-rustls-comp"
] }
#deadpool-redis = { version = "0.15.1", features = ["rt_tokio_1"] }
```

```rust
use axum::extract::{Request, State};
use axum::http::{StatusCode};
use axum::response::{Response};
use axum::routing::get;
use axum::{middleware, Router};
use std::time::Duration;
use axum::middleware::Next;
use tokio::net::TcpListener;
use tower_http::timeout::TimeoutLayer;
use chrono::Utc;
use redis::Client;
// use deadpool_redis::{Config, Connection, Pool, Runtime};

// pub type RedisClientPool = Pool;

#[tokio::main]
async fn main() -> anyhow::Result<()> {  
    let connection_uri = "rediss://localhost:6379";
    let redis_client = redis::Client::open(connection_uri)?;
    
    // let cfg = Config::from_url(connection_uri);
    // #[allow(clippy::expect_used)]
    // let redis_pool = cfg
    //     .create_pool(Some(Runtime::Tokio1))
    //     .expect("unable to create redis client");
    
    let timeout_router = Router::new()
        .route("/blub", get(handle))
        .with_state(redis_client)
        .layer(TimeoutLayer::new(Duration::from_millis(100)))
        .layer(middleware::from_fn(stats));

    let router = Router::new().nest("/application", timeout_router);

    let listener = TcpListener::bind("127.0.0.1:5000").await?;
    println!("Starting application on {listener:?}");
    axum::serve(listener, router).await?;
    Ok(())
}

pub async fn stats(request: Request, next: Next) -> Response {
    let start = Utc::now();
    let response = next.run(request).await;
    let after_request = Utc::now();
    let response_time = after_request.timestamp_millis() - start.timestamp_millis();

    println!("{} {}: {}ms", start, after_request, response_time);
    
    response
}

pub async fn handle(State(redis): State<Client>) -> Result<String, StatusCode> {
    let result = String::from("data");

    // let connection: Connection = redis.get().await.unwrap(); 
    let connection = redis.get_multiplexed_async_connection().await.unwrap();
    
    tokio::time::sleep(Duration::from_millis(500)).await;

    Ok(result)
}
```

Some facts I want to highlight. As you can see, there is a small web server setup. As we were investigate our issues, we created this application to test. Therefore, we have a controller with a timeout layer. It's basically a select! on the handler and a sleep feature. Which one is finished first will return. In my real application monitoring I saw quite high p99s, a lot higher than our controller timeout. A bit suspicious.  
We also have a layer (`stats`) to measure the request time. The handler is simple. We open a redis connection, waiting 500 ms and return some data. The sleep is non-blocking. 
The expected result is, a timeout (408) after 100ms when requesting `http://localhost:5000/application/blub`. But it did not work. The response time is between 210ms - 300ms in my tests. 

Now, let's remove this line `let connection = redis.get_multiplexed_async_connection().await.unwrap();`. What do we see? A timeout after 102ms. Yeah, we excepted this. Just out of curiosity, change the `tokio::time::sleep(Duration::from_millis(500)).await` to `std::thread::sleep(Duration::from_millis(500))`. Also, 500ms, which are more than the 100ms timeout. This is due to [std::thread::sleep](https://doc.rust-lang.org/std/thread/fn.sleep.html) is blocking and does not yield.  
But the issue looks identically. Something is blocking in this line: `let conn = redis.get_multiplexed_async_connection().await.unwrap();`. Btw. I also tried deadpool-redis on top of redis. Same results. 

We tried to fix the issue. A few ideas:

- Switching the connection uri to `redis://localhost:6379` will resolve the issue. (rediss to redis). 
- Switching from `tokio-rustls-comp` to `tokio-native-tls-comp` resolves this also.

Okay, both are not really options for us. Company policies require the use of SSL/TLS for the communication. And also, rustls is preferred over native-tls due to easier building and overall less headache. So @FalkWoldmann noticed, that we should try out to enable the `tls-rustls-webpki-roots` feature next to `tokio-rustls-comp` in the redis create. And yes. **It also fixes our issue**. 

We think maybe the call to the **system's PKI root storage is blocking**, but we did not have the time to validate our thoughts. [webpki](https://github.com/rustls/webpki-roots) contains Mozilla's root certificates and should be enough to validate our certificates. We do not work with self-signed certificates. 

After this fix, our application can handle 4–5 times the requests on an even smaller instance. A great win. 

**tl;dr**: webpki-roots are nice and non-blocking. native-pki-roots are blocking. Why? I don't know. I do not have the skills to dig deeper into it. It's might be an issue on tokio-rustls, or somewhere else. But anyhow, we should talk about setting webpki-roots as the default.

**Comment by jaymell** [comment]
> In any case, it sounds like this might be relevant more to the rustls repo, if you think the issue is with their implementation.

Yeah, I expect so. You're likely hitting slow Disk IO on loading the native certs. IIUC the web-pki certs are effectively bundled in your binary, so less of an issue there. As nihohit said, this is likely an upstream issue.
